### PR TITLE
Refactor/zcash

### PR DESCRIPTION
**Pull request by collinc97**

### CODE REVIEW

**Comment by collinc97** [comment]
The current build fails due to an error in the bitcoin and zcash tests.

File: zcash/src/address.rs [hunk 1]
```diff
@@ -1,67 +1,103 @@
...
 use serde::Serialize;
-use sha2::{Digest, Sha256};
 use std::fmt;
 
+/// Represents the format of a Zcash address
+#[derive(Serialize, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Format {
+
+    /// Unshielded Zcash Address
+    Unshielded,
```
Comment (amousa11) [inline zcash/src/address.rs]: They call these `transparent`

**Comment by amousa11** [comment]
I think it might be easier to review all of these changes if you broke it up into a few PRs. One for zcash changes, one for eth, one for btc changes

**Comment by amousa11** [comment]
many of the eth changes are covered/made obsolete by #13

File: zcash/src/address.rs [hunk 2]
```diff
@@ -1,67 +1,103 @@
...
+            Some((Format::Shielded, network)) => Self::shielded(public_key, &network),
+            None => Self::unshielded(public_key, &Network::Mainnet)
+        }
+    }
+}
+
+impl ZcashAddress {
+
+    /// Returns an unshielded address from a given Zcash public key
+    fn unshielded(public_key: &ZcashPublicKey, network: &Network) -> Self {
```
Comment (amousa11) [inline zcash/src/address.rs]: transparent

File: bitcoin/src/wallet.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
+use address::{Address, Type};
```
Comment (amousa11) [inline bitcoin/src/wallet.rs]: not sure if we want to use the wallet anymore, it may be worth deleting for now

File: ethereum/src/address.rs [hunk 4]
```diff
@@ -44,3 +44,39 @@ impl fmt::Display for Address {
         write!(f, "{}", self.address)
     }
 }
+
```
Comment (amousa11) [inline ethereum/src/address.rs]: #13 addresses this, i'd remove this

File: ethereum/src/keypair.rs [hunk 5]
```diff
@@ -106,3 +106,95 @@ impl fmt::Display for KeyPair {
         write!(f, "{}", self.private_key)
     }
 }
+
```
Comment (amousa11) [inline ethereum/src/keypair.rs]: #13 is dealing with these changes as well

File: ethereum/src/wallet.rs [hunk 6]
```diff
@@ -23,7 +23,7 @@ impl EthereumWallet {
         EthereumWallet { keypair, address }
     }
 
-    /// Recovers a EthereumWallet from a KeyPair object
+    /// Recovers an EthereumWallet from a KeyPair object
```
Comment (amousa11) [inline ethereum/src/wallet.rs]: #13 deletes the wallet as well

File: zcash/src/address.rs [hunk 7]
```diff
@@ -1,67 +1,103 @@
...
-                .to_vec(),
+impl Address for ZcashAddress{
+    type Format = (Format, Network);
+    type PrivateKey = ZcashPrivateKey;
+    type PublicKey = ZcashPublicKey;
+
+    /// Returns the address corresponding to the given Zcash private key.
+    fn from_private_key(private_key: &Self::PrivateKey, format: Option<Self::Format>) -> Self {
+        match format {
+            Some((Format::Unshielded, _)) => Self::unshielded(&private_key.to_public_key(), &private_key.network),
```
Comment (amousa11) [inline zcash/src/address.rs]: transparent

File: zcash/src/lib.rs [hunk 8]
```diff
@@ -12,18 +12,22 @@ extern crate serde;
...
 
-pub mod utils;
-pub use self::utils::*;
+pub mod public_key;
+pub use self::public_key::*;
 
-pub mod wallet;
-pub use self::wallet::*;
\ No newline at end of file
+//pub mod utils;
```
Comment (amousa11) [inline zcash/src/lib.rs]: if `pub mod utils` is commented out you probably aren't using `utils` so you can just delete it from this lib

File: zcash/src/private_key.rs [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+        test_new(private_key);
+    }
+
+    #[test]
+    fn test_new_testnet() {
+        let private_key = ZcashPrivateKey::new(Network::Testnet);
+        test_new(private_key);
+    }
+
+//    #[test]
```
Comment (amousa11) [inline zcash/src/private_key.rs]: any reason these are commented out?

**Pull request closed without merging**